src/main.rs

```diff
@@ -39,12 +39,15 @@ fn main() -> CliResult {
 fn run_floki_from_args(args: &Cli) -> Result<(), Error> {
     debug!("Got command line arguments: {:?}", &args);
 
-    let environ = environment::Environment::gather(&args.config_file)?;
-    debug!("Got environment {:?}", &environ);
+    let environ = || -> Result<environment::Environment, Error> {
```

Comment:
Unsure if this is more legible than a function... I'm undecided.

Am wondering if a separate function that just returned a tuple of (environment, config) might be easier to reason about.

Reply:
Returning a tuple:
- it entangles the two things that don't need to be together,
- from experience, it makes it a pain to refactor.

"Legibility vs a function"?
- is that concern driven by an unfamiliarity with closures?  It's almost exactly what we written before but with a `||` put in front.

Separate function?
- How come there wasn't already a separate function?
  + Because all that would be doing is make these two one liner calls and returning a tuple.
  + That would then be polluting the top level namespace with a function we only want to call in this scope.
  + It would require the reader to do more jumping around the code to find out how the env and config are got.
- I'd say all these reason how true now those one liners to fetch env and config have been made closures.
- If the processes to get these data was really involved, I'd say make it a separate function (or functions), but they're one liners, I don't think there's any useful abstractions to make over them.

Reply:
In an earlier version of `floki`, possibly an entirely parallel entity I maintained myself at home before it was open sourced, the environment and config structures were munged into one core structure which held all configuration. It depends on how you look at it, they are gathered differently, but accessed largely together (access patterns ought to be a stronger driver of grouping) - most functions which take one take the other unless what they are doing is very specific. So it's not clear they belong apart, necessarily. I'm not sure it would be terrible to fetch them together, but I'm also not sure it's necessary, or at least, necessitated by this change.

Max is plenty well acquainted with closures :-).

To be honest, my first hunch was just to inline the calls in the match statement, which is essentially what the closures do (in essence, both closures are just shorter names for stuff which already exists). Sure, we need to pass `args.config_file` by hand, but that's no biggy.

What about the logging? Well, for some reason, autism or something, I never like to put logging all over the place, so I tend to keep it in the top level functions. This is purely me being stupid, because some dark corner of my mind hates logging. I've not even been consistent in this respect. Both of the functions here are already I/O heavy; the log about the environment may as well go in the `gather` function, similarly the log which fetches the config file mutatis mutandis. Not only is that appropriate, it's probably better.

Then we're left with two functions that do what we need already, and can just move the calls to the relevant match arms.

Reply:
I'd say there's good reason to want to keep logging at the highest level function that you can.

Part of the point of a function is to be able to call it more than once.  As soon as you do, what's the say that a log statement made by the function should be the same severity in all cases? You also have to scrub any context from the log message that might not apply to ever time the function is called.  This can be "OK" within an application, but in a library it's pretty hard generally to make assertions about when, why, and in what context it's called.  Something to think about for whoever implements #112 anyway.

For now I'll go with your suggestion of dropping the log statements down into the functions the first level down, which as you say, is consistent with other parts of floki.